File: tests/benchmarks/aot_c90/runtime/string_concat_builder.c
/* C reference for string_concat_builder.xr: each `s + t` allocates a fresh
 * string of the combined length (same allocation pattern as the Xray
 * implementation, which rebuilds the string per concatenation). */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static char *concat2(const char *a, size_t la, const char *b, size_t lb, size_t *out_len) {
    char *r = (char *) malloc(la + lb + 1);
    if (!r)
        abort();
    memcpy(r, a, la);
    memcpy(r + la, b, lb);
    r[la + lb] = 0;
    *out_len = la + lb;
    return r;
}

static int64_t run(int64_t rounds, int64_t parts) {
    int64_t total = 0;
    for (int64_t r = 0; r < rounds; r++) {
        char *s = (char *) malloc(1);
        size_t slen = 0;
        s[0] = 0;
        for (int64_t i = 0; i < parts; i++) {
            char piece[32];
            int plen = snprintf(piece, sizeof(piece), "part-%lld;", (long long) (i % 10));
            size_t nlen;
            char *ns = concat2(s, slen, piece, (size_t) plen, &nlen);
            free(s);
            s = ns;
            slen = nlen;
        }
        total += (int64_t) slen;
        free(s);
    }
    return total;
}

int main(void) {
    printf("%lld\n", (long long) run(4000, 40));
    return 0;
}

File: tests/benchmarks/aot_c90/runtime/string_concat_builder.rs
// Rust reference for string_concat_builder.xr (safe; mirrors the Xray
// rebuild-per-concat pattern: each `s + t` produces a fresh String).

fn run(rounds: i64, parts: i64) -> i64 {
    var mut total: i64 = 0;
    var mut r: i64 = 0;
    while r < rounds {
        var mut s = String::new();
        var mut i: i64 = 0;
        while i < parts {
            s = s + "part-" + &(i % 10).to_string() + ";";
            i += 1;
        }
        total += s.len() as i64;
        r += 1;
    }
    total
}

fn main() {
    println!("{}", run(4000, 40));
}
